[docs] update docs for intro to dev builds (#35400)
expo/expo: docs/pages/develop/development-builds/introduction.mdx

The **JavaScript bundle (`npx expo start`)** is where your app's UI code and business logic is. In a production apps there is one `main.js` bundle that is shipped with the app itself. In development this JS bundle is live reloaded from your local machine. The main role of React Native is to provide a way for the JavaScript code to access the native APIs (Image, Camera, Notifications etc). But only APIs and libraries that were bundled in the **native app** can be used.

Since the native app bundle is immutable, this means that when you're using Expo Go for development, you can only rely on the native code and tools that exist in Expo Go. The only way to get around it is to build your native app yourself instead of using our pre-packaged sandbox. This is exactly what a **Development Build is: your own version of Expo Go** where you are free to use any native libraries and change any native config.

<VideoBoxLink
  videoId="7J8LRpja9_o"

>   title="Getting the most out of Expo Development Builds"

Reviewer: ```suggestion
  title="Getting the most out of Expo Development builds"
```

Author: Removed

  description="In this talk Kadi explains the history of how Expo Go came to be, it's benefits and limitations."
/>

## What can't you do in Expo Go and why

> Migrating from Expo Go to a Development Build can be a bit overwhelming the first time. Expo Go is the perfect tool for learning, prototyping and experimenting, but most production apps will convert to using Development builds sooner rather than later. It helps to know exactly what is _not_ possible in Expo Go and _why_, so you can make an informed decision on when and why to make this move.

Reviewer: ```suggestion
Migrating from Expo Go to a development build can be a bit overwhelming the first time. Expo Go is the perfect tool for learning, prototyping, and experimenting, but most production apps will convert to using Development builds sooner rather than later. It helps to know exactly what is _impossible_ in Expo Go and _why_, so you can make an informed decision on when and why to make this move.
```

Just a thought: I think the first line should be more of a positive message. This sentence implies that migration is going to be hard, which might not be a good first impression considering the type of the audience you described in our last conversation. Right now, it seems that the developer is being forced to use a development build instead of Expo Go.

Author: Sure, I'll just remove the first sentence.

<VideoBoxLink
  videoId="FdjczjkwQKE"
  title="Expo Go & Development Builds: which should you use?"

>   description="In this tutorial video Beto explains what each of them is, and when to choose a Development Build."

Reviewer: ```suggestion
  description="In this tutorial video Beto explains what each of them is, and when to choose a development build."
```

/>

### Use libraries with native code that aren't in Expo Go

> Take [react-native-webview](https://docs.expo.dev/versions/latest/sdk/webview/) for example. This library contains native code, but it [is included in Expo Go](https://github.com/expo/expo/blob/main/apps/expo-go/package.json#L23). So when you `npx expo install react-native-webview` on your project, it will install the library in your `node_modules` which includes both the JS code and the native code. But the JS bundle you are building _only_ uses the JS code. Then your JS bundle gets uploaded to Expo Go, and it interacts with the native code that was already bundled with the app.

Reviewer: ```suggestion
Consider [`react-native-webview`](/versions/latest/sdk/webview/) as an example, a library that contains native code, but [is included in Expo Go](https://github.com/expo/expo/blob/main/apps/expo-go/package.json#L23). When you run `npx expo install react-native-webview` command in your project, it will install the library in your **node_modules** directory, which includes both the JS code and the native code. But the JS bundle you are building _only_ uses the JS code. Then, your JS bundle gets uploaded to Expo Go, and it interacts with the native code that was already bundled with the app.
```

> If instead you tried to use a library that is not included, e.g. [react-native-firebase](https://docs.expo.dev/guides/using-firebase/#using-react-native-firebase), then you can use the JS code and hot reload the new bundle into Expo Go but it will immediately error because the JS code is trying to call into the native code from the React Native Firebase package that does not exist in Expo Go. There is no way to get the native code into the Expo Go app unless it was already included in the bundle that was uploaded to the App / Play Store.

Reviewer: ```suggestion
Instead, when you try to use a library that is not included, for example, [`react-native-firebase`](/guides/using-firebase/#using-react-native-firebase), then you can use the JS code and hot reload the new bundle into Expo Go but it will immediately error because the JS code tries to call the native code from the React Native Firebase package that does not exist in Expo Go. There is no way to get the native code into the Expo Go app unless it was already included in the bundle that was uploaded to the app stores.
```

### Test changes in app icon, name, splash screen

> If you're developing your app in Expo Go only, you _can_ build a store version of it that will use your provided values and images for there, it just won't be possible to test it in Expo Go.

Reviewer: ```suggestion
If you're developing your app in Expo Go only, you can build a store version that will use your provided values and images; it just won't be possible to test it in Expo Go.
```

> All of these are native assets that get shipped with the native bundle and are immutable once installed. The Expo Go app does show a "splash screen which" which is just your app icon on a solid color background. This is a dev-only emulation of a splash screen we added so you could see what the splash screen will probably look like. It is limited however, e.g. you cannot test `SplashScreen.setOptions` to animate the splash screen.

Reviewer: ```suggestion
These native assets are shipped with the native bundle and are immutable once the app is installed. The Expo Go app does show a splash screen, which is your app icon on a solid color background. This is a dev-only emulation to view how the splash screen will probably look. However, it is limited, for example, you cannot test `SplashScreen.setOptions` to animate the splash screen.
```

### Remote push notifications

> While in-app [notifications](https://docs.expo.dev/versions/latest/sdk/notifications/) are available in Expo Go, remote push notifications (that is, sending a push notification from a server into the app) are not. This is because a push notification service should be tied to your own push notifications certificates and while it is possible to make it work in Expo Go, it often caused confusion for production builds. It is therefore better to test remote push notifications in Development Builds so you can ensure parity in behavior between development and production.

Reviewer: ```suggestion
While [in-app notifications](https://docs.expo.dev/versions/latest/sdk/notifications/) are available in Expo Go, remote push notifications (that is, sending a push notification from a server to the app) are not. This is because a push notification service should be tied to your own push notification certificates, and while it is possible to make it work in Expo Go, it often causes confusion for production builds. It is recommneded to test remote push notifications in development builds so you can ensure parity in behavior between development and production.
```

-"...while it is possible to make it work in Expo Go, it often caused confusion for production builds." I think you can omit this since SDK 53, push notifications will be completely unavailable.

### Universal Links / App Links

> Both [iOS Universal Links](https://docs.expo.dev/linking/ios-universal-links/) and [Android App Links](https://docs.expo.dev/linking/android-app-links/) require a two-way association between the native app and the website. In particular, it requires for the native app ro include the URL for the lined website. This is not possible with Expo Go due to the aforementioned native code immutability.

Reviewer: ```suggestion
Both [Android App Links](/linking/android-app-links/) and [iOS Universal Links](/linking/ios-universal-links/) require a two-way association between the native app and the website. In particular, it requires the native app to include the linked website's URL. This is impossible with Expo Go due to the aforementioned native code immutability.
```

- I think you meant "linked" website. I am not sure.

> ### (iOS Device only) open projects using older SDKs

Reviewer: ```suggestion
### (iOS device only) open projects using older SDKs
```

> Expo Go can only support one SDK version at a time. So, when we release a new SDK version we need to update Expo Go to support the new SDK version and this will be the only version of Expo Go available to install from the stores.

Reviewer: ```suggestion
Expo Go can only support one SDK version at a time. When a new SDK version is released, Expo Go is updated to support the newer version, and this will be the only version of Expo Go available to install from the stores.
```

> If you're developing on Android Device, Android Emulator or iOS Simulator, a compatible version of Expo Go can be [downloaded and installed](https://expo.dev/go). The only platform where this is not possible is iPhone devices because apple does not support side-loading older versions of apps.

Reviewer: ```suggestion
If you're developing on an Android Device, Android Emulator, or iOS Simulator, a compatible version of Expo Go can be [downloaded and installed](https://expo.dev/go). The only platform where this is impossible is iPhone devices because Apple does not support side-loading older versions of apps.
```

> ## How to convert from Expo Go to a Development build

Reviewer: ```suggestion
## How to convert from Expo Go to a development build
```

To migrate from Expo Go, you'll need to

> 1. install `expo-dev-client`
> 2. build the native app using native build tools

Reviewer: ```suggestion
1. Install `expo-dev-client`
2. Build the native app using native build tools
```

> The details on how to do this based on the platform you're building for is covered in the [next section](/develop/development-builds/create-a-build).

Reviewer: ```suggestion
The details on how to build the native app are platform-dependent and are covered in the [next guide](/develop/development-builds/create-a-build).
```

> The Expo Dev Client package includes the launcher UI (shown in the screenshots below), dev menu, extensions to test updates etc. The Expo Go app has the dev menu built in, so we'll need to install it ourselves for a development build.

Reviewer: ```suggestion
The Expo Dev Client library includes the launcher UI (shown in the screenshots below), dev menu, extensions to test updates, and more. The Expo Go app has the dev menu built in, so we'll need to install it ourselves for a development build.
```

It's unclear as to what "so we'll need to install it ourselves for a development build." the last sentence means.

Author: Hm I'm not sure how to phrase it better. Expo Go had the dev tools built in. With dev builds we're no longer using Expo Go so we need to install the dev tools separately. Maybe changing it to "separately" instead of "ourselves" helps?

Reviewer: ah okay, I wasn't aware of that. Yeah, changing the phrase might be more informative, like: "so you'll need to install it separately for a development build.". I'll leave this as a separate suggestion on the changes you've already implemented.

<ContentSpotlight

>   alt="expo-dev-client launcher and menu UIs on Android and iOS"

Reviewer: ```suggestion
  alt="expo-dev-client launcher and menu UIs on Android and iOS."
```

### Universal Links / App Links

Both [Android App Links](/linking/android-app-links/) and [iOS Universal Links](/linking/ios-universal-links/) require a two-way association between the native app and the website. In particular, it requires the native app to include the linked website's URL. This is impossible with Expo Go due to the aforementioned native code immutability.

### (iOS device only) open projects using older SDKs

Expo Go can only support one SDK version at a time. When a new SDK version is released, Expo Go is updated to support the newer version, and this will be the only version of Expo Go available to install from the stores.

If you're developing on an Android Device, Android Emulator, or iOS Simulator, a compatible version of Expo Go can be [downloaded and installed](https://expo.dev/go). The only platform where this is impossible is iPhone devices because Apple does not support side-loading older versions of apps.

## How to convert from Expo Go to a development build

To migrate from Expo Go to a development build, you'll need to:

<Step label="1">

### Install the `expo-dev-client`

> The Expo Dev Client library includes the launcher UI (shown in the screenshots below), dev menu, extensions to test updates, and more. The Expo Go app has the dev menu built in, we'll need to install it separately for a development build.

Reviewer: ```suggestion
The Expo Dev Client library includes the launcher UI (shown in the screenshots below), dev menu, extensions to test updates, and more. The Expo Go app has the dev menu built in, and that's why you need to install it separately for a development build.
```

I think adding the clarity that installing the Expo Dev Client library is bringing in dev tools is good to note here. Otherwise, some folks might get confused as to where they need to install these dev tools 😄
